spec.md

```diff
@@ -133,7 +133,44 @@ If the blob is not found in the registry, the response code MUST be `404 Not Fou
 
 #### Push
 Pushing an image works in the opposite order as a pull: the blobs making up the layers are pushed first, and the
-manifest last.
+manifest last. A useful diagram is provided [here](https://github.com/google/go-containerregistry/tree/d7f8d06c87ed209507dd5f2d723267fe35b38a9f/pkg/v1/remote#anatomy-of-an-image-upload).
+
+> What is the order of a push for some client?
+
+Generally, the data dependency between components drives the order of operations.
+For example, you can't upload a mainfest without knowing the config blob and layer digests. You can't know the config blob digest without knowing the layer diffids. You can't know the layer digests until you've gzipped and uploaded them (unless you know them ahead of time). This generally means that we do an upload like:
+
+ - blobs are uploaded first (config and layers)
+ - blobs (optionally) are then linked to an image manifest (and the manifest uploaded)
+ - the layers and manifest are linked to a repository
+
+Keep in mind that along with image layers, the config is a [separate blob](https://github.com/opencontainers/image-spec/blob/master/config.md#example) that describes the container runtime specification.
+
+> How do I design my registry models for Blobs, Image Manifests, and Repositories?
+
+By way of specifying the name of the repository (`<name>`) for upload endpoints below,
+this gives you freedom to design your registry in several ways:
+
+ - Having blobs linked 1:1 to a repository, and then referenced in Image Manifests only belonging to that repository (you wouldn't need to worry about blob sharing across repositories, and cleanup would be limited to ensuring the blob only isn't needed by the repository in question)
+ - Having blobs linked to Image Manifests, which are linked to Repositories.
+ - Having blobs not linked to any Image Manifest or Repository and shared amongst all blobs (and for this implementation idea you would need to ensure that blobs are not deleted that are still being used, and that there aren't any security loopholes with someone uploading a blob that would be used by another repository.
+
+If there is another design idea not listed, please contribute it to the specification. For all of the above designs, you should consider storage (organization), along with permissions and cleanup. Different registries with varying user groups will likely have different use cases that warrant different behavior.
+
+> How do I handle garbage collection?
+
+Based on the design you choose above, you might want to implement integrity checking. For example, when you upload a manifest all blobs that it references should already exist in your registry. 
```

Comment:
This all seems verbose and not strictly related to push, which I think warrants an FAQ section